File: server/models/Transaction.model.js
import mongoose from 'mongoose';
import { v4 as uuidv4 } from 'uuid';

const transactionSchema = new mongoose.Schema({
  transactionId: { 
    type: String, 
    unique: true, 
    default: () => 'TXN-' + uuidv4().toUpperCase().substring(0, 12)
  },
  sender: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  beneficiary: { type: mongoose.Schema.Types.ObjectId, ref: 'Beneficiary', required: true },
  
  // Amount details
  sendAmount: { type: Number, required: true },
  sendCurrency: { type: String, required: true, default: 'USD' },
  receiveAmount: { type: Number, required: true },
  receiveCurrency: { type: String, required: true },
  exchangeRate: { type: Number, required: true },
  transferFee: { type: Number, required: true, default: 0 },
  totalDeducted: { type: Number, required: true },
  
  // Payout method
  payoutMethod: { 
    type: String, 
    enum: ['bank_deposit', 'cash_pickup', 'mobile_wallet', 'card'],
    required: true 
  },
  
  // Payment method (how sender pays)
  paymentMethod: {
    type: String,
    enum: ['bank_transfer', 'debit_card', 'credit_card'],
    required: true
  },
  
  // Status tracking
  status: { 
    type: String, 
    enum: ['pending', 'processing', 'sent', 'delivered', 'failed', 'cancelled', 'refunded'],
    default: 'pending'
  },
  statusHistory: [{
    status: String,
    timestamp: { type: Date, default: Date.now },
    note: String,
    updatedBy: String
  }],
  
  // Countries
  sendCountry: { type: String, required: true },
  receiveCountry: { type: String, required: true },
  
  // Purpose
  transferPurpose: { type: String, enum: ['family_support', 'education', 'medical', 'business', 'gift', 'other'], default: 'family_support' },
  
  // Compliance & fraud
  isFlagged: { type: Boolean, default: false },
  flaggedReason: { type: String },
  flaggedAt: { type: Date },
  reviewedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  reviewedAt: { type: Date },
  reviewNote: { type: String },
  
  // Reference numbers
  bankReference: { type: String },
  providerReference: { type: String },
  
  // Timestamps
  initiatedAt: { type: Date, default: Date.now },
  completedAt: { type: Date },
  estimatedDelivery: { type: Date },
  
  // IP and device info for fraud detection
  ipAddress: { type: String },
  deviceInfo: { type: String },
  
  // Notes
  senderNote: { type: String, maxlength: 200 },
}, { timestamps: true });

// Index for performance
transactionSchema.index({ sender: 1, createdAt: -1 });
transactionSchema.index({ transactionId: 1 });
transactionSchema.index({ status: 1 });
transactionSchema.index({ isFlagged: 1 });

export default mongoose.model('Transaction', transactionSchema);

File: transfastApp/lib/features/transactions/models/transaction_model.dart
// lib/features/transactions/models/transaction_model.dart

class TransactionModel {
  final String id;
  final String transactionId;
  final double sendAmount;
  final String sendCurrency;
  final double receiveAmount;
  final String receiveCurrency;
  final double exchangeRate;
  final double transferFee;
  final double totalDeducted;
  final String payoutMethod;
  final String paymentMethod;
  final String status;
  final String sendCountry;
  final String receiveCountry;
  final String transferPurpose;
  final bool isFlagged;
  final String? flaggedReason;
  final String? senderNote;
  final String? estimatedDelivery;
  final String? completedAt;
  final String createdAt;
  final Map<String, dynamic>? beneficiary;
  final List<Map<String, dynamic>> statusHistory;

  const TransactionModel({
    required this.id,
    required this.transactionId,
    required this.sendAmount,
    required this.sendCurrency,
    required this.receiveAmount,
    required this.receiveCurrency,
    required this.exchangeRate,
    required this.transferFee,
    required this.totalDeducted,
    required this.payoutMethod,
    required this.paymentMethod,
    required this.status,
    required this.sendCountry,
    required this.receiveCountry,
    required this.transferPurpose,
    required this.isFlagged,
    this.flaggedReason,
    this.senderNote,
    this.estimatedDelivery,
    this.completedAt,
    required this.createdAt,
    this.beneficiary,
    required this.statusHistory,
  });

  factory TransactionModel.fromJson(Map<String, dynamic> json) {
    return TransactionModel(
      id: json['_id'] ?? '',
      transactionId: json['transactionId'] ?? '',
      sendAmount: (json['sendAmount'] as num?)?.toDouble() ?? 0,
      sendCurrency: json['sendCurrency'] ?? 'USD',
      receiveAmount: (json['receiveAmount'] as num?)?.toDouble() ?? 0,
      receiveCurrency: json['receiveCurrency'] ?? '',
      exchangeRate: (json['exchangeRate'] as num?)?.toDouble() ?? 0,
      transferFee: (json['transferFee'] as num?)?.toDouble() ?? 0,
      totalDeducted: (json['totalDeducted'] as num?)?.toDouble() ?? 0,
      payoutMethod: json['payoutMethod'] ?? '',
      paymentMethod: json['paymentMethod'] ?? '',
      status: json['status'] ?? 'pending',
      sendCountry: json['sendCountry'] ?? '',
      receiveCountry: json['receiveCountry'] ?? '',
      transferPurpose: json['transferPurpose'] ?? '',
      isFlagged: json['isFlagged'] ?? false,
      flaggedReason: json['flaggedReason'],
      senderNote: json['senderNote'],
      estimatedDelivery: json['estimatedDelivery'],
      completedAt: json['completedAt'],
      createdAt: json['createdAt'] ?? '',
      beneficiary: json['beneficiary'] as Map<String, dynamic>?,
      statusHistory: (json['statusHistory'] as List?)
              ?.cast<Map<String, dynamic>>() ?? [],
    );
  }

  String get beneficiaryName {
    if (beneficiary == null) return 'Unknown';
    return '${beneficiary!['firstName'] ?? ''} ${beneficiary!['lastName'] ?? ''}'.trim();
  }
}

// Beneficiary model
class BeneficiaryModel {
  final String id;
  final String firstName;
  final String lastName;
  final String country;
  final String currency;
  final String phone;
  final String relationship;
  final String payoutMethod;
  final Map<String, dynamic>? bankDetails;
  final Map<String, dynamic>? walletDetails;
  final int totalTransfers;

  const BeneficiaryModel({
    required this.id,
    required this.firstName,
    required this.lastName,
    required this.country,
    required this.currency,
    required this.phone,
    required this.relationship,
    required this.payoutMethod,
    this.bankDetails,
    this.walletDetails,
    required this.totalTransfers,
  });

  String get fullName => '$firstName $lastName';
  String get initials => '${firstName[0]}${lastName[0]}'.toUpperCase();

  factory BeneficiaryModel.fromJson(Map<String, dynamic> json) {
    return BeneficiaryModel(
      id: json['_id'] ?? '',
      firstName: json['firstName'] ?? '',
      lastName: json['lastName'] ?? '',
      country: json['country'] ?? '',
      currency: json['currency'] ?? '',
      phone: json['phone'] ?? '',
      relationship: json['relationship'] ?? 'family',
      payoutMethod: json['payoutMethod'] ?? 'bank_deposit',
      bankDetails: json['bankDetails'] as Map<String, dynamic>?,
      walletDetails: json['walletDetails'] as Map<String, dynamic>?,
      totalTransfers: json['totalTransfers'] ?? 0,
    );
  }
}
